torchtune/models/t5/_encoder.py

```diff
@@ -0,0 +1,304 @@
+        rel_pos_max_dist (int): Maximum distance for relative positions.
+            Distances beyond this are grouped into the last bucket.
+            See: :class:`~torchtune.models.t5._encoder.T5EncoderRelativePositionBias`
+        max_seq_len (int): The maximum sequence length (context length) of the model.
+    """
+
+    def __init__(
+        self,
+        *,
+        token_embedding: nn.Embedding,
+        layer: nn.Module,
+        final_norm: nn.Module,
+        num_layers: int,
+        num_heads: int,
+        rel_pos_num_buckets: int,
+        rel_pos_max_dist: int,
+        max_seq_len: int,
+    ):
+        super().__init__()
+        self.token_embedding = token_embedding
+        self.layers = nn.ModuleList([copy.deepcopy(layer) for i in range(num_layers)])
+        self.final_norm = final_norm
+        self.max_seq_len = max_seq_len
+        self.relative_position_bias = T5EncoderRelativePositionBias(
+            num_buckets=rel_pos_num_buckets,
+            max_dist=rel_pos_max_dist,
+            num_heads=num_heads,
+            max_seq_len=max_seq_len,
+        )
+
+    def forward(self, tokens: Tensor) -> Tensor:
+        """
+        Args:
+            tokens (Tensor): input tensor with shape ``[bsz, max_seq_len]``
+
+        Returns:
+            Tensor: output tensor with shape [bsz, max_seq_len, embed_dim]
+
+        Raises:
+            ValueError: if seq_len of tokens is bigger than max_seq_len
+        """
+        # Input validation
+        bsz, seq_len = tokens.shape
+        if seq_len > self.max_seq_len:
+            raise ValueError(
+                f"seq_len ({seq_len}) of input tensor should be smaller "
+                f"than max_seq_len ({self.max_seq_len})"
+            )
+
+        # Input embedding [bsz, max_seq_len] -> [bsz, max_seq_len, embed_dim]
+        x = self.token_embedding(tokens)
+
+        # Bias added to the attention scores of every layer (to add relative position information)
+        rel_pos_bias = self.relative_position_bias()
+
+        # Encoder
+        for layer in self.layers:
+            x = layer(x, rel_pos_bias)
+
+        return self.final_norm(x)
+
+
+class T5EncoderLayer(nn.Module):
+    """
+    Single layer of the T5 encoder (standard transformer layer with relative position bias).
+
+    Args:
+        attn (MultiHeadAttention): Attention module.
+        mlp (nn.Module): Feed-forward module.
+        sa_norm (nn.Module): Normalization to be applied before self-attention.
+        mlp_norm (nn.Module): Normalization to be applied before the feed-forward layer.
+    """
+
+    def __init__(
+        self,
+        attn: MultiHeadAttention,
+        mlp: nn.Module,
+        sa_norm: nn.Module,
+        mlp_norm: nn.Module,
+    ) -> None:
+        super().__init__()
+        self.attn = attn
+        self.mlp = mlp
+        self.sa_norm = sa_norm
+        self.mlp_norm = mlp_norm
+
+    def forward(self, x: Tensor, rel_pos_bias: Tensor) -> Tensor:
```

Comment:
Well, kind of... it's a float tensor that gets added to the attention scores (which is how `scaled_dot_product_attention` deals with float masks, so yeah we could think of it as a mask). 

I didn't use our modules because: 
- we can't use flex attention with float masks
- our `MultiHeadAttention`/`TransformerSelfAttentionLayer` modules expect the masks to be boolean tensors (according to the docstring at least)
- `MultiHeadAttention` uses the default attention scaling (`1/sqrt(dim)`), but T5 doesn't scale it at all
- `MultiHeadAttention` uses `F.scaled_dot_product_attention` which is much slower with float masks than the manual implementation I went with

We could switch to our modules with a couple small changes (update the docstrings to clarify that the mask can also be a float tensor and add an argument for disabling attention scaling), but given how little code is required to just implement separate versions for T5, I thought it was cleaner to leave our modules alone (especially since this implementation is faster).